libs/agno/agno/tools/parallel.py

```diff
@@ -0,0 +1,300 @@
+except ImportError:
+    raise ImportError("`parallel-web` not installed. Please install using `pip install parallel-web`")
+
+
+class CustomJSONEncoder(json.JSONEncoder):
+    """Custom JSON encoder that handles non-serializable types by converting them to strings."""
+
+    def default(self, obj):
+        try:
+            return super().default(obj)
+        except TypeError:
+            return str(obj)
+
+
+class ParallelTools(Toolkit):
+    """
+    ParallelTools provides access to Parallel's web search and extraction APIs.
+
+    Parallel offers powerful APIs optimized for AI agents:
+    - Search API: AI-optimized web search that returns relevant excerpts tailored for LLMs
+    - Extract API: Extract content from specific URLs in clean markdown format, handling JavaScript-heavy pages and PDFs
+
+    Args:
+        api_key (Optional[str]): Parallel API key. If not provided, will use PARALLEL_API_KEY environment variable.
+        enable_search (bool): Enable Search API functionality. Default is True.
+        enable_extract (bool): Enable Extract API functionality. Default is True.
+        all (bool): Enable all tools. Overrides individual flags when True. Default is False.
+        max_results (int): Default maximum number of results for search operations. Default is 10.
+        max_chars_per_result (int): Default maximum characters per result for search operations. Default is 10000.
+        beta_version (str): Beta API version header. Default is "search-extract-2025-10-10".
+        mode (Optional[str]): Default search mode. Options: "one-shot" or "agentic". Default is None.
+        include_domains (Optional[List[str]]): Default domains to restrict results to. Default is None.
+        exclude_domains (Optional[List[str]]): Default domains to exclude from results. Default is None.
+        max_age_seconds (Optional[int]): Default cache age threshold (minimum 600). Default is None.
+        timeout_seconds (Optional[float]): Default timeout for content retrieval. Default is None.
+        disable_cache_fallback (Optional[bool]): Default cache fallback behavior. Default is None.
+    """
+
+    def __init__(
+        self,
+        api_key: Optional[str] = None,
+        enable_search: bool = True,
+        enable_extract: bool = True,
+        all: bool = False,
+        max_results: int = 10,
+        max_chars_per_result: int = 10000,
+        beta_version: str = "search-extract-2025-10-10",
+        mode: Optional[str] = None,
+        include_domains: Optional[List[str]] = None,
+        exclude_domains: Optional[List[str]] = None,
+        max_age_seconds: Optional[int] = None,
+        timeout_seconds: Optional[float] = None,
+        disable_cache_fallback: Optional[bool] = None,
+        **kwargs,
+    ):
+        self.api_key: Optional[str] = api_key or getenv("PARALLEL_API_KEY")
+        if not self.api_key:
+            log_error("PARALLEL_API_KEY not set. Please set the PARALLEL_API_KEY environment variable.")
```

Comment:
We raise in other tools when missing critical keys. Let's raise a `ValueError` here too